### ISSUE TITLE
workflow #2 - qubes-doc-community-fork

### ISSUE THREAD

**Issue by adubois** [Qubes-Community/Contents#4]
The aim of this issue is to test and discuss the idea of forking qubes-doc to act as a staging area for it as well as an exploration area for new subjects/topics to be documented.

[edit taradiddles - renamed issue title, hope you don't mind]

**Comment by adubois** [comment]
I am going to explore this: https://help.github.com/articles/fork-a-repo/
In particular I want to test the Keep your fork synced and assess if:
- it is feasible with a community repo
- once the command is typed, do you need a client CLI to execute the commands every time you want to sync, or is the web UI of GitHub changed to allow to do so via this interface

**Comment by adubois** [comment]
OK, so created a blank repo.
Then in a disposable terminal:

~~~
git clone https://github.com/Qubes-Community/qubes-doc-community-fork.git
cd qubes-doc-community-fork
git remote -v
git remote add upstream https://github.com/QubesOS/qubes-doc.git
git remote -v
git fetch upstream
git merge upstream/master
git push origin master

**Comment by adubois** [comment]
There are few use cases to document:
- Contribution (branch, edit, commit, PR to Qubes Community, merge, PR to Qubes)... I'm not sure how the end bit works, particularly if we want to do a PR that match a merge on community side.
- Fetch (change in upstream, who to merge it in Qubes Community)... I think the process above is what needs to be done. We need to create a dom0 script that automate the sequence. I'll have to wait for a change upstream to test this.

**Comment by Aekez** [comment]
this is looking great, I'm looking forward to see it all working.

btw if we want to make local changes and not full repository changes, then how about this? I memorized it from some guide somewhere, so I'm not sure if it's correct way to do it, but I've used it to make local edits in an repo rather than try push/merge the whole full repository.

After git cloning whole repository or pull to update existing one, login etc.
- change the needed files/sub-folders.
- cd to base dir containing the files or sub-folders that were changed. 
- while in base dir of changes, do 'git init' (I'm not sure if this is needed every time, I think it is important to reach the repository config files?).
- while in base dir of changes, do 'git status' (all changes are in red).
- while in base dir of changes, do 'git add .' 
- while in base dir of changes, do 'git status' (all changes are now green).
- while in base dir of changes, do 'git commit' (might be faster to add '-m message' and maybe also -S for local path?, I didn't use -S though).
- while in base dir of changes, do 'git push origin master' (or available sub-branch).
- while in base dir of changes, do 'git merge' (merge with master if desired/allowed).

I haven't fully found out how it works yet with this approach, but it has worked nonetheless, but I gotta iron out any potential mistakes, so I will continue reading. For example I didn't use the git checkout in there, I wonder if there are any dangers by not making or using a different branch as a master, i.e. it might be better to use branch first for files history, even if one has master access, or?

**Comment by adubois** [comment]
I found this guide very brief and with the right spirit (you need to branch so that while you change things, someone else can change other things and you can have individual PR).

https://sourceforge.net/p/mingw-w64/wiki2/git/

If I remember the step "working on your changes" can be summarized in one command (I think git checkout -b MyFeatureBranch but you get less output feedback)
Second point to note when you do git status the UI tells you what to do (to move forward with your change or revert).
When we document in the repo qubes-community how to contribute, this guide should be referenced and we make our own copy taking into account our local context.

If you work on multiple changes, you need to remember to checkout to the branch that match your change (so you group all the modification in one PR and don't mix things up).

I think teaching people how to use git rather than using wiki a valuable for the community.

Git is useful for any work you do. You can have your local repo and manage your local word documents or whatever you do anywhere, keeping the unit of work small in separate branches, and merging changes as you progress.

**Comment by Aekez** [comment]
> If you work on multiple changes, you need to remember to checkout to the branch that match your change (so you group all the modification in one PR and don't mix things up).

aha! so that's the purpose of checkout. Thanks for clarifying that one. I'm one step closer to grasping it now. I assume if having shifted checkout on multiple branches, then "git push origin branch" at the end might mess it up if it tries to execute multiple of checkout branches into a single origin branch? So this is here where one might make a mistake with it?

> I think teaching people how to use git rather than using wiki a valuable for the community.

Absolutely agreed, it's a great tool, I can already feel it a bit from the little exposure I had to it, albeit I'm still ways off mastering it just yet. Gonna take a few days to get comfy with the basics/normal use.

**Comment by taradiddles** [comment]
>> I think teaching people how to use git rather than using wiki a valuable for the community.

Yes, hopefully once someone contributes to the wiki and sees how PRs are created, he/she would learn basic git usage to submit his/her next contribution. However git shouldn't be a hard requirement for contributing docs to this project -  many people simply don't have the time to learn how to use a new tool.

> I can already feel it a bit from the little exposure I had to it, albeit I'm still ways off mastering it just yet. Gonna take a few days to get comfy with the basics/normal use.

... which confirms what I just wrote :)

@adubois : branches may avoid conflicts when automatically sync'ing the repo with upstream, but is there a way to create branches automatically in a way that is totally transparent to contributors ?
* if no, then IMHO we'd better have people copy/paste from the original qubes-doc repo, because that's an order of magnitude easier to do than creating branches.
* if yes, how would branches/commits will be merged later on ? Or are they supposed to be deleted once the PR gets included in the official qubes-doc or in the qubes-community repositories ?

**Comment by adubois** [comment]
>@adubois : branches may avoid conflicts when automatically sync'ing the repo with upstream, but is there a way to create branches automatically in a way that is totally transparent to contributors ?
>    if no, then IMHO we'd better have people copy/paste from the original qubes-doc repo, because that's an order of magnitude easier to do than creating branches.
 >   if yes, how would branches/commits will be merged later on ? Or are they supposed to be deleted once the PR gets included in the official qubes-doc or in the qubes-community repositories ?

@taradiddles The answer yes. You just click on the .md file to view it, click on the pencil to edit it. I have protected the branch master, so the only option is to commit in a new branch.
So if you use the web interface, there is nothing to do for the new user.
As a reviewer we have to look at how to best accept it:
- Accept and merge in community then do PR in qubes
or
- accept and do PR in Qubes, then merge.

I suspect the second option is the right on. Because anyway if Qubes reject the PR, we may want to store the info in another area of the documentation.

I have not used the wiki on Github, maybe it is based on Git. Which repo is a wiki here so I can look into the functionalities?

**Comment by taradiddles** [comment]
> You just click on the .md file to view it, click on the pencil to edit it

I've created a dummy github user ('test-q'), forked the repo (a branch named 'patch-1' was created), changed a file and made a PR, but the PR doesn't show up in qubes-doc-community-fork, maybe because my new account was flagged for no obvious reason (I tried to edit the wiki twice, got 404 errors, and the 'your account is flagged' banner appeared).

Anyway, my PR shows  `test-q  wants to merge 1 commit into Qubes-Community:master from test-q:patch-1` (the repo name is missing, it is 'qubes-doc-community-fork').
Is that how it's supposed to be ? I'd have thought that there would be a new branch in the qubes-doc-community-fork to avoid conflicts with qubes-os' repo, not a new branch in my forked repo - but I'm a total git noob so maybe that's normal.

> I have not used the wiki on Github, maybe it is based on Git. Which repo is a wiki here so I can look into the functionalities?

IIRC you can [edit wikis with git](https://stackoverflow.com/questions/38816161/editing-a-wiki-on-a-github-repo#44360202). Qubes-Community/Qubes-Community's wiki is [here](https://github.com/Qubes-Community/Qubes-Community/wiki).

**Comment by Aekez** [comment]
>... which confirms what I just wrote :)

Yeps :)

> I have protected the branch master, so the only option is to commit in a new branch.

That probably explains my banging head against the keyboard just now "oh man, why the heck did it stop working!" x) I'll get into the habit of sending to sub-branches from now on, it's probably a good habit to have anyway I imagine, even if owners?
 

> Is that how it's supposed to be ? I'd have thought that there would be a new branch in the qubes-doc-community-fork to avoid conflicts with qubes-os' repo, not a new branch in my forked repo

Good question.. I get the gut feeling there can be an official one and an unofficial one though? maybe the official one sticks around, but a user can make their own push branch if desired? I wonder if that is correctly understood though, it is a guess after all.

**Comment by adubois** [comment]
I have only protected the master branch of qubes-doc-community-fork. So if you have problems elsewhere let me know.
I have put 2 teams as validator of the PR.

**Comment by Aekez** [comment]
I see :) then it must have been something else, I'll have to try find out what I did different. It might be some hours till I get around to it though, I will keep getting major distractions today, so I'll be slow unfortunately. If I'm lucky I can find some good clues to go on or the very reason, with little time, though I'll have to see how it goes. Right now I can't tell you what I did wrong as I have no clues to go on yet (except perhaps the missing checkout that I haven't implemented in my routine yet but was planning to).

**Comment by adubois** [comment]
Sent from my mobile phone.

> On 10 Mar 2018, at 13:05, Yuraeitha <notifications@github.com> wrote:
> 
> If you work on multiple changes, you need to remember to checkout to the branch that match your change (so you group all the modification in one PR and don't mix things up).
> 
> aha! so that's the purpose of checkout. Thanks for clarifying that one. I'm one step closer to grasping it now. I assume if having shifted checkout on multiple branches, then "git push origin branch" at the end might mess it up if it tries to execute multiple of checkout branches into a single origin branch? So this is here where one might make a mistake with it?
> 
> I think teaching people how to use git rather than using wiki a valuable for the community.
> 
> Absolutely agreed, it's a great tool, I can already feel it a bit from the little exposure I had to it, albeit I'm still ways off mastering it just yet. Gonna take a few days to get comfy with the basics/normal use.
> 

I know 2-3 workflows, but it is sufficient

**Comment by adubois** [comment]
>> I have put 2 teams as validator of the PR.

>That's great :) Two tiers of moderator access level I assume?

There was 2 teams "quality proof reading" and "quality security" if I remember well. I put both but I am not sure if we need approval from a person from each team or if any user from any of the teams is OK. We'll see.

**Comment by adubois** [comment]
OK, I've tested the wiki.
you have versioning (= git)
you don't have PR (>git: don't have to wait for validation, but you can fork if you want, but <git as you can't quality check before it is final)
you can't reference users with @ (< git)
you can't reference issues with # (<git)
you can't fetch upstream changes from qubes (<git)
you can't PR directly with qubes (<git also this needs to be confirmed)
you can't automatically generate a web-site (and merge sites between repos) (<git)

So the only benefit I see is the scratch area, but this is the exact purpose of a branch, you create a branch do all the change you want, when you're done (or fed up) you raise a PR (or delete the branch)

I probably have missed some points. Please list the plus you see. For me at this point, we should forbid wiki, it going to be cancer. But this is at this point and if I can't convince you, I am fine to keep it (but it is a mistake :smiley: )

**Comment by awokd** [comment]
Thanks for testing that out, @adubois . The main plus I see to the wiki is the lower barrier of entry to non-Github familiar people ("what's a branch?" would be their first question) but I'm concerned there will be a lot of admin overhead/cleanup involved too.

**Comment by adubois** [comment]
We can try both. I started a doc on how to contribute. It is simple. As most contributors will have had a fair amount of exposure to IT,  a concise how-to with the 2-3 most common work-flows will allow 95% of people to feel OK with git.
By the way, if you don't do anything complicated (i.e. just edit a page), you don't need to know about branch, because when you press the edit button, the equivalent of the save button (and the only option) is create a branch, and the next button is create a pull request (with help on both screen on why).
We'll see... and adapt to the best method by iterating. However as soon as we feel one way or another does not work, we need to make sure we correct. I don't want a back log of 500 stupid PR, or a messy wiki with duplicates with qubes-doc, etc...

**Comment by adubois** [comment]
Sent from my mobile phone.

> On 10 Mar 2018, at 14:50, Ivan <notifications@github.com> wrote:
> 
> You just click on the .md file to view it, click on the pencil to edit it
> 
> I've created a dummy github user ('test-q'), forked the repo (a branch named 'patch-1' was created), changed a file and made a PR, but the PR doesn't show up in qubes-doc-community-fork, maybe because my new account was flagged for no obvious reason (I tried to edit the wiki twice, got 404 errors, and the 'your account is flagged' banner appeared).
> 
> Anyway, my PR shows test-q wants to merge 1 commit into Qubes-Community:master from test-q:patch-1 (the repo name is missing, it is 'qubes-doc-community-fork').
> Is that how it's supposed to be ? I'd have thought that there would be a new branch in the qubes-doc-community-fork to avoid conflicts with qubes-os' repo, not a new branch in my forked repo - but I'm a total git noob so maybe that's normal.
> 
If you do a simple file edit, you don’t need to create a branch in your fork. You just browse Qubes community and click the edit button

**Comment by Aekez** [comment]
Great work with the permissions and reflection perspectives :) 

I had to sleep first, I'll start looking into the git issues now and see if I can find out why it didn't work. 

By the way about a training/testing grounds, we have two different approaches as far as I can see. One is to tell people to train on their own github, another is to have a training/testing grounds here Qubes Community. Telling people to train on their own github page might not always work for some people (people can work in mysterious ways), but having a training/testing grounds here on Qubes Community does have it pros as well, for example we can easier see what people did wrong and help them, rather than having to go to their page to look. But it would also introduce yet another repository though, which is the downside, and another downside is extra work for us, is our goal to teach? I think you guys like teaching as well, but it'll be hard for us if we try to do too much.

Another plus side though, is if we make step by step guides, and tell them to do it on the training ground repository, then we can see if they for example put the files in the right folders as the specific guide told them to do. For example we can make branch switch traps in the guides, where the guide is not wrong, but the trap is in the nature of not understanding it correctly, means it will lead to making a visible mistake in the training grounds. Failure is a great positive way to learn.

But it's also a question how much time we would spend teaching others, I mean, we spend a lot of time ourselves before asking any questions. Some people work the other way around and ask questions before doing anything themselves. Maybe we will see in time how all this turns out?

**Comment by Aekez** [comment]
btw in qubes-doc-community-fork repository, how is the current idea to putting together drafts and review them? For example if an audit is made to existing qubes doc's, but it requires complex edits, or in general it would be edits in the Qubes Community's name, since it's through the fork from here right? 

So if anyone can go in and do it through there, it would be an edit from that individual, but in the Qubes Community name, without the Qubes Community having been involved in that edit. Is this how it works? Do we have a means to protect against something like this? I mean if an edit of Qubes docs is done, or publishing is done, in Qubes Community's name, without involving the Qubes Community, then wouldn't it be better if they did it through their own account?

This is all new to me, so I might be concerned about something that isn't a problem here. Do we have the means to keep taps on for example a Qubes doc branch which is a branch we made, and only allow people to make changes in Qubes Community's name, to that branch, and that branch alone?

So for example, if people don't use that branch, then it won't be in the Qubes Community's name. But keeping it to a single branch, we can better and more easier control rogue community commits. Is that possible?

**Comment by awokd** [comment]
I'm OK with links to existing content somewhere explaining Github, but don't think we should have to maintain a training ground for people.
Commits against our fork with have the commiter's name on them. If they look good then we do a PR against qubes-doc and the same commits carry across, if I understand the flow right.

**Comment by Aekez** [comment]
agreed, so while we can support guides or questions people might have, we should ask people to put in effort themselves to learn. Seems more reasonable in my view as well, if they want to help contribute they must put in some effort themselves.

Permissions was already done to the teams, but I changed it a bit though by segmenting the team's purpose https://github.com/orgs/Qubes-Community/teams/quality-review-teams/teams for example, a team can review any other Qubes community repositories, but they can't push to the Qubes official doc's. Then another team, specializes purely in pushing to Qubes official doc's.

Also we owners don't need teams, we can do anything anyway. But it gives some official look, and we can add none-owners to the teams as well. I'm going to create teams for the other projects and do something similar to them. So that the main branch is protected until merges are considered.

What about @one7two99 [799] I haven't seen him post here yet, is he aware of these discussions? They are kind of hidden away in a repository after all.

**Comment by Aekez** [comment]
> Commits against our fork with have the commiter's name on them. If they look good then we do a PR against qubes-doc and the same commits carry across, if I understand the flow right.

By the same commits carrying across, you mean the trace of who did what? credits for work done etc.? So even if only a team can push to the Qubes official doc's, will whoever made the work still get the traceable credits?

**Comment by awokd** [comment]
> credits for work done etc

I believe so, since it's a fork. Only exception might be for content manually copied & pasted out of the wiki into a commit.

**Comment by Aekez** [comment]
> I believe so, since it's a fork. Only exception might be for content manually copied & pasted out of the wiki into a commit.

That's amazing, it removes my worry. GitHub truly is an interesting work-system. I also agree the wiki copy/paste issue seems troublesome, maybe we can do manual edits and add authors to solve it? We're already moving text manually at that point anyway, adding authors shouldn't be much extra work.

https://github.com/orgs/Qubes-Community/teams/community-teams/discussions
This message was mostly only a test, feel free to edit it btw.

**Comment by one7two99** [comment]
Hello guys,

Am 11.03.2018 3:20 nachm. schrieb "Yuraeitha" <notifications@github.com>:

What about @one7two99 <https://github.com/one7two99> [799] I haven't seen
him post here yet, is he aware of these discussions? They are kind of
hidden away in a repository after all.

I'm reading in silent mode, and thought that you've enough skills to figure
something out and if it doesn't work we evolve from there ;-)

I need to migrate my local docs to the new Qubes Community Docs in the next
days.

[799]

**Comment by Aekez** [comment]
@one7two99 good to see you're here! :)
Feel free to add your thoughts and opinions though, that's one of the central cores this place is all about right after all. At least when it comes to how I perceive it, but I think everyone sees it that way.

About the Project teams, I added this message 
https://github.com/orgs/Qubes-Community/teams/project-teams/discussions
It's just to get started, we can edit/add/remove the message to be more appropriate. Does everyone agree that Projects should retain some levels of independence? or should we turn it around and make them all open to full Community? The projects I added btw are fully open to everyone here. I just imagine some future members might want to retain more control.

btw what should we call the administrators team? doing @administrators each time might be a bit long. Should we re-name it to something more community like? Using the @owners flag seem a bit overkill too if the purpose is to create a community that feels closer to equal and open.

**Comment by taradiddles** [comment]
> As most contributors will have had a fair amount of exposure to IT

I know people who use Qubes (eg. journalists) and they're far from having a "fair amount of exposure to IT". Granted, they most likely represent a minority of users.

> So the only benefit I see is the scratch area, but this is the exact purpose of a branch, you create a branch do all the change you want, when you're done (or fed up) you raise a PR (or delete the branch)

See, that's the problem: technically and/or conceptually, branches - or anything requiring git proficiency - make it *harder* for git noobs or "non-git" users to submit PRs here than submitting PRs *directly* to the official qubes-doc repo. This goes contrary to a main goal of this project: **ease** contributions to qubes-doc.

> I probably have missed some points. Please list the plus you see. For me at this point, we should forbid wiki, it going to be cancer. But this is at this point and if I can't convince you, I am fine to keep it (but it is a mistake smiley )

The only plus is that it's not git, and that's a big plus. Wikipedia wouldn't be what it is now if people had to learn how to use github and submit PRs. Maybe nobody will ever use the wiki, maybe it will be an administrative nightmare and we'll nuke it, but let's see what happens first.

I don't think I have the required git knowledge to manage PRs in the project ; I'll be more comfortable not being an owner. I'll still help with getting PRs in shape and transforming scratch wiki content into PRs, as well as adding stuff to qubes-community with PRs, as a standard user.

I don't think I've seen a fully documented contribution test in the comments so far, we should try to make a few and see how it works out. IMHO we're focusing too much on github's functionality and complex cases.
Re- synchronization script: we don't have to wait for qubes-doc to be modified, we can fork another repo where we can change things and check that it works well with branches when the same files are edited in both repos.
(sry if it's already done - didn't see it).

**Comment by Aekez** [comment]
@taradiddles 
> I don't think I have the required git knowledge to manage PRs in the project ; I'll be more comfortable not being an owner. 

If it ever comes to that, then you can still be an owner without being on the push team. Of course you can still push as all owners can do that irregardless to being a member of a team or not, but it won't be official that way, and if you don't push then it won't be in the logs either. Does that remove worries?
Our git knowledge should also be increasing pretty quickly if we manage it daily or occasionally if busy with other things in life. Becoming good at git doesn't seem so hard if a bit persistent with it, learn a thing or two once in a while, by all appearances it can be mastered somewhat quickly without learning coding languages etc. But if it takes away worries (I can relate, I feel the same about other things), you can at any time remove yourself from the push team if desired, and still be an owner. You don't have to step down for that, imho :)

**About the wiki**
I agree we can just test it out and see how it goes (I very much like the idea of a wiki as well). It seems like it has good potential, of course, with potential downsides that we need to be careful about. But if we look in the qubes-users mail thread and use that as an example, in general there are many friendly people in there, most are not flaming, spamming or creating issues. At most it's lacking posts with not enough information, but that is okay for a wiki I suppose, others can fill in the gaps in a wiki fashion, it's what wiki's are all about anyway. But in general, the kind of people that find their way to Qubes OS, just seem much less destructive, than compared to people from other parts of the internet. Maybe the type of people that find their way to Qubes OS will change in the future, but right now it doesn't seem like many would cause problems, many of the current Qubes user generation appear responsible. I agree that we can always re-visit a wiki discussion if it becomes a problem in the future.

Ivan do you want to try set up the wiki you have in mind? I can do it if you're busy, I just didn't do it before because I wasn't sure if you wanted to do it or had some thoughts in mind about how you wanted to set it up.

**Comment by Aekez** [comment]
@Qubes-Community/administrators 
The Qubes staff has been discussing and reached a conclusion to let us be independent. So we're now from this point on-wards officially an independent volunteer community. 

The feeling I got from their response is that they're positive about what we're doing here, at least for now. They may come looking one day, as normal members only. But it's important to remember such visits will be unofficial, they will have no ties back to The Qubes OS Project while members around here. 

They did not say this explicitly, but I think it's good to underline (stressing the point) that we should respect their choice and freedom to become normal users, so that they can come and go as they feel like, just like any other normal member. 

I forwarded that they, as The Qubes OS Project, can come with requests to us if they feel something needs changing, I hope that's okay by everyone here? I have not gotten a response back on this particular point however, but the way I interpret this, is that if any such request should ever happen, it'll be by official channels from The Qubes OS Project. However, in contrast if they come around here as members, they will be normal members only. The distinction here between these two perspectives is important to remember.

**Comment by awokd** [comment]
> independent

I'm going to take that as a vote of confidence. :) We seem to have a good mix of skills and enough professionalism to not mess it up too badly. Besides, that's how I wrote up the disclaimer so it saves me the trouble of having to change it! "This site is run by volunteers. The Qubes OS Project is not affiliated with this site and does not endorse the content of any of these pages."

**Comment by Aekez** [comment]
> I'm going to take that as a vote of confidence. :) 

Kind of like a bird pushed out of its nest to learn to fly (funny mental picture of falling down, * cough *) well, we might need some more training before we can fly properly :)

> We seem to have a good mix of skills and enough professionalism to not mess it up too badly.  

Agreed, and if we do screw up, then we just have to recover and learn from it. It's the best learner in life anyway, not daring to make mistakes only results in slow learning. So if wanting the most success then we have to be bold, even if it can be a bit scary. This project has already given me one sleepless night as it is, but there is nothing but to push on-wards if we want to succeed with this project.

> Besides, that's how I wrote up the disclaimer so it saves me the trouble of having to change it! "This site is run by volunteers. The Qubes OS Project is not affiliated with this site and does not endorse the content of any of these pages." 

yeah I noticed that too, I had you in mind when I read that :)


I've put forward a new discussion topic in the issue tracker on building a strategic statement, hopefully it can help us clear out what we are, and what we want to become. It makes it not only easier for ourselves, but also easier for anyone who comes looking trying to find out what we're all about. It's much more simple than it looks btw, so feel free to bring up suggestions without worries of all that theoretical foundation it rests on. It's a lot to read up on for something that is actually very simple; to define ourselves, define what we want to become, and define how we want to archive it; one does not need to know more than that to start a discussion on it.

**Comment by Aekez** [comment]
It should probably be noted that the "no ties back to The Qubes OS Project" when here as members, is my interpretation. Now that I think about it, saying "only as members" can be understood in two different ways. Either "just as members" means a semi-cut or a full cut, my interpretation was a full cut, but it can just as well mean a semi-cut. For example just as members can mean that they retain their ties backwards to The Qubes OS Project, but is still just members here. This largely comes down to how to interpret it, we should probably keep this in mind too. I'm not aware of which one of these they prefer, but it should be fine to leave that up to them to decide. This interpretation difference shoulnd't change much how we run things here though, since they still want us to be independent.

**Comment by Aekez** [comment]
I moved these threads to an appropriate channel where they belong due to included private information in them, the transfer is documented. Apologies for the inconvenience.

**Comment by adubois** [comment]
Apologies, I really can't find the time to read all the text above... I would could try to be as concise as possible (when you want me to contribute) that would help.

OK back on the topic of this issue: qubes fork and understanding how it works.
I spent 5min at lunch time to ask @awokd to add me as reviewer for his PR.
After he did, I could then approve it, which unlocked it (as a review is mandatory for branch master).
The problem is that I merged it in master, we probably don't want that until we know it is approved upstream... I am not sure. If the option is present, I should have merged in community branch.
It feels to me that having another branch with all the community changes is better, and that master is a reflection of upstream (Qubes OS).
But I have not come across this type of usage of git. So I'll need to do some research tonight.

I suspect that I'll have to delete the repo and set-it up again with:
- master and community branch.
- community being the default, master being a reflection of upstream.
- master and community will need review before approving PR.
- We should never have to merge a PR in master (as Qubes will upstream)

But I am not sure that it will be this way.

If you find good article on maintaining git project that are a fork of another project, while maintaining/accepting upstream changes, let me know.

**Comment by awokd** [comment]
If we don't merge to our master at some point, how do we upstream "approved" changes? Flip side is how do we handle changes that get rejected for merges to qubes-doc?

**Comment by adubois** [comment]
I found this blog post: https://www.atlassian.com/git/articles/git-forks-and-upstreams
So it seems that is to fork, you never really merge into master. You create pull request from a branch.
- if accepted, you get is by fetching
- if rejected, I think we should decide to merge into a community branch (which is a fork of our master).

This way, by looking at branch master you have Qubes doc, and by looking at branch community, you have all our upstream rejected change. Doing a diff between the too you have all the additional bits provided by the community. We may be able to generate a page that list these automatically...
@Qubes-Community/administrators may I ask you to go to the qubes-staging-fork repo and review+approve the changes there.
Thanks.

**Comment by adubois** [comment]
Maybe this is a more useful read... https://stackoverflow.com/questions/17630970/how-to-manage-the-open-source-projects-patches-which-cant-push-to-upstream
The last answer seems pretty simple. The only thing I may want to add is to merge the patch in the community branch if I know it is going to be rejected by upstream....

**Comment by Aekez** [comment]
Below is a solution suggestion, but it has one major issue in order to work, but I think it's just one issue, instead of many issues, so if we can solve that issue, then this layout might work. 

First we might want to unpin the official Qubes doc's though, since if we want to ensure no one pushes in the Qubes Community branch name, then only that branch will be allowed in that forked repository anyway, and that means only a select few people will do that, while the rest don't need to worry about the official repository. 

Instead we could pin the repository that the Community use to make changes or create new doc's. I.e. a "shadow" of the forked official repository. A shadow, (a word I made up btw but you get the meaning as it's quite literal to the meaning of real life shadow), is a direct cast-off the forked repository. So if any changes is made in the official repository (changes from the Qubes staff), then this change will follow down into the fork, as well as the shadow. 

**Original repository <---->** **Forked repository** <----> **Shadow repository** 

The fork is different because it's a portal between Qubes staff / Qubes Community. While the shadow repository is different because it's a portal between Qubes Community members / Qubes community publisher push team. Both the forked and shadow repository maintain current live changes done upstream by the Qubes staff, however, any changes made in the shadow repository will not be uploaded upstream, instead it remains in the shadow repository. A publisher team can then move finished shadow content out of shadow, and into the forked repository, where the Qubes staff then will see the request, to which they either accept or decline. The forked repository is therefore only a transition repository, while the shadow repository is the work repository.

So if we pin the shadow repository (or whatever we call it), and then unpin the fork to the Qubes staff's official Qubes doc's, then the members only need to worry about the shadow one, and not the forked one. Only the publish push team needs to be aware of both of the repositories. 

However is a shadow repository possible? I.e. is it possible to make a repository which is updated from the original owners live, while we work on it ourselves without publishing to the original owners? This is the only thing I see that hinders this kind of system, I don't know if a shadow like repository is possible or not. I.e. the shadow repository should update from upstream, but it must not be allowed to update upstream to the original owners, or to the forked repository without the publisher teams action to do so.

**Comment by Aekez** [comment]
Though another issue is that we need to keep the shadow repository clean, so that clutter doesn't build up over time. I'm not sure how easy or hard that will be, or if it becomes a nightmare to keep it in sync.

**Comment by Aekez** [comment]
If making a clone, rather than a fork, i.e. making the shadow repository a clone, and the transition repository a fork, would the clone repository then still be able to obtain live changes from the original one? I know clones can be cut off fully, but can they be made to still receive updates from the original too?

**Comment by taradiddles** [comment]
@adubois: any ETA ? If you're too busy and/or synchronization becomes too complicated maybe we should revert to simple copy/paste for the time being, until you have time to test/polish this workflow.

**Comment by awokd** [comment]
I'm hoping to get some time this weekend to start digging into these workflows too.

**Comment by Aekez** [comment]
It might be working to just have a single branch if all Qubes doc's will go through a singular repository issue system though. All other branches should be locked off, when coming from the Community github? If anyone wants to make Qubes doc PR on their own, it can still be done from their own github account?

**Comment by awokd** [comment]
Yes, that's what the part about

>I set up branch protection on master so it should also require a PR. I created a branch off master called community, but I'm not sure this is needed- I expect people will create their own branches for edits.

means. Not sure we need the community branch.

**Comment by awokd** [comment]
Not exactly; see revisions in my comment. Try submitting and approving your own PR against our qubes-doc/master, you should get a better idea of how it works.

**Comment by Aekez** [comment]
I'll give it a shot on a testing repository outside the community first, copying the repository permission structure, and trial and error'ing a bit in the sandbox. It's a good suggestion to get my hands dirty on the branches/pull requests, should be a good way to go about it.

**Comment by awokd** [comment]
@adubois What do you think about the process in https://github.com/Qubes-Community/Contents/issues/4#issuecomment-373912753? Keeping it in sync with upstream looks to be the most annoying part since it's a manual process.

**Comment by awokd** [comment]
Looking at @adubois 's https://github.com/Qubes-Community/Contents/issues/4#issuecomment-372463775 again, I can see the intent of the community branch but it might result in merge conflicts. Assuming the sync process would be:

qubescomm/comm -> qubesos/master -> qubescomm/master -> back to qubescomm/comm

As I see it, differences that grow between qubesos/master and qubescomm/comm could eventually lead to merge conflicts.

Propose simplifying this workflow by having one forked repo and branch for documents intended for merge, and backing out commits that get rejected upstream (wish there was a better way to handle this, though). Not finding any way to automatically sync from upstream, so it would have to be done by script or manually. Weekly might suffice.

**Comment by taradiddles** [comment]
\+ 1 for simplifying ; I can't really help with that git stuff but I can host the sync script(s) and set a cronjob on one of my public always-on servers (provided the script runs unattented - should be the case for git --reset or similar stuff).

once you set this up I'll update the 'contributing' doc with the branch's name.

**Comment by awokd** [comment]
Deleted the qubes-doc-community-fork repo; let's use qubes-doc repo going forward. Will document here the sync commands.

**Comment by awokd** [comment]
Initial setup
~~~
git config --global user.email "emailaddress"
git config --global user.name "githubusername"
~~~
Sync
~~~
git clone https://github.com/Qubes-Community/qubes-doc.git
cd qubes-doc
git remote add upstream https://github.com/QubesOS/qubes-doc.git
git checkout master
git pull https://github.com/QubesOS/qubes-doc.git master
git push origin master
~~~
The last two commands are interactive; need to figure out how to automate them. I'll plan on doing this procedure weekly until then.

**Comment by awokd** [comment]
Closing for now, this should be workable.
